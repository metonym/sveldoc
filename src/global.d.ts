declare module "prettier-plugin-svelte";
declare module "prism-svelte";
