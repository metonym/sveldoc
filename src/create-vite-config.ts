import { typescript } from "svelte-preprocess";
import { svelte } from "@sveltejs/vite-plugin-svelte";
import type { UserConfig } from "vite";
import { getTsconfigPaths } from "./utils/get-tsconfig-paths";
import { pluginIndex } from "./plugin-index";
import { pluginReadme } from "./plugin-readme";
import { preprocessReadme } from "./preprocess-readme";

type CreateViteConfig = (options?: UserConfig) => UserConfig & {
  test: {
    globals: boolean;
    environment?: string;
  };
};

export const createViteConfig: CreateViteConfig = (options) => {
  return {
    ...options,
    resolve: {
      ...options?.resolve,
      alias: {
        ...options?.resolve?.alias,
        ...getTsconfigPaths(),
      },
    },
    plugins: [
      pluginIndex(),
      svelte({
        extensions: [".svelte", ".md"],
        preprocess: [typescript(), preprocessReadme()],
      }),
      pluginReadme(),
    ],
    test: {
      globals: true,
      environment: "jsdom",
    },
    optimizeDeps: {
      exclude: ["prism-svelte"],
    },
  };
};
