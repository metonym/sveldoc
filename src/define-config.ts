import { typescript } from "svelte-preprocess";
import { svelte } from "@sveltejs/vite-plugin-svelte";
import type { UserConfig } from "vite";
import { getTsconfigPaths } from "./utils/get-tsconfig-paths";
import { pluginIndex } from "./plugin-index";
import { pluginReadme } from "./plugin-readme";
import { preprocessReadme } from "./preprocess-readme";

export interface DefineConfigOptions extends UserConfig {
  /**
   * Specify additional CSS styles to be
   * injected into the main index file.
   */
  styles?: string;

  /**
   * Set to `true` to not apply default
   * GitHub Markdown styles to iframes.
   * @default false
   */
  resetStyles?: boolean;

  /**
   * Specify the default branch used to
   * permalink relative URLs in the README.
   * @default "master"
   */
  branch?: string;
}

type DefineConfig = (options?: DefineConfigOptions) => UserConfig & {
  test: {
    globals: boolean;
    environment?: string;
  };
};

export const defineConfig: DefineConfig = (options) => {
  const resetStyles = options?.resetStyles === true;
  const base = options?.base;
  const branch = options?.branch;

  return {
    ...options,
    base,
    resolve: {
      ...options?.resolve,
      alias: {
        ...options?.resolve?.alias,
        ...getTsconfigPaths(),
      },
    },
    plugins: [
      pluginIndex({
        resetStyles,
      }),
      svelte({
        extensions: [".svelte", ".md"],
        preprocess: [typescript(), preprocessReadme({ base, branch })],
      }),
      pluginReadme(),
    ],
    test: {
      globals: true,
      environment: "jsdom",
    },
  };
};
