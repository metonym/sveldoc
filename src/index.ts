export { createViteConfig } from "./create-vite-config";
