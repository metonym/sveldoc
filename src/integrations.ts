import { readFileSync, writeFileSync } from "fs";
import { transformReadme } from "./transform-readme";
import { match } from "./utils";
import type { PluginOption } from "vite";
import type { PreprocessorGroup } from "svelte/types/compiler/preprocess";

/**
 * Svelte preprocessor that transforms
 * content in a `README.md` file.
 */
export const preprocessReadme: () => Pick<PreprocessorGroup, "markup"> = () => ({
  markup: ({ content: source, filename }) => {
    if (!filename || !match.readmeFile(filename)) return;
    return transformReadme({ source, filename });
  },
});

/**
 * Vite plugin that transforms the `README.md`
 * but omits adding Svelte stuff.
 *
 * Only executes when building the app.
 */
export const pluginReadme: () => PluginOption = () => {
  let filename: null | string = null;

  return {
    name: "vite-plugin-readme",
    apply: "build",
    load(id) {
      if (match.readmeFile(id)) {
        filename = id;
        return undefined;
      }
    },
    writeBundle() {
      if (filename)
        writeFileSync(
          filename,
          transformReadme({
            source: readFileSync(filename, "utf-8"),
            filename,
            noEval: true,
          }).code
        );
    },
  };
};
