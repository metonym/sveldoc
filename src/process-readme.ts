import fs from "fs";
import path from "path";
import { marked } from "marked";
import prettier from "prettier";
import "prettier-plugin-svelte";
import Prism from "prismjs";
import "prism-svelte";
import { extractComponentPath } from "./utils/extract-component-path";
import { match } from "./utils/match";
import { parseComponent } from "./utils/parse-component";
import { getPackageJson } from "./utils/get-package-json";

const package_json = getPackageJson();

marked.use({
  walkTokens: (token: any) => {
    if (
      token.type === "link" &&
      /^\.\//.test(token.href) &&
      package_json?.repository?.url
    ) {
      token.href = path.join(
        package_json.repository.url,
        "tree/master",
        token.href
      );
    }
  },
});

interface ProcessReadmeOptions {
  source: string;
  filename: string;
  noEval?: boolean;
}

export const processReadme = async (options: ProcessReadmeOptions) => {
  const { source, filename, noEval } = options;

  let cleaned = "";
  let open = false;

  source.split("\n").forEach((line, index, lines) => {
    if (!open) {
      cleaned += line + "\n";
      open = match.exampleStart(line);
    }

    if (match.exampleEnd(lines[index + 1])) open = false;
  });

  let script_module_unique_lines = new Set();
  let script_unique_lines = new Set();
  let dependencies = new Set<string>();
  let styles = "";
  let lines_code = "";

  for await (let line of cleaned.split("\n")) {
    lines_code += (noEval ? line : marked.parse(line)) + "\n";

    if (match.exampleStart(line)) {
      const path_component = extractComponentPath(line);

      if (!path_component) continue;
      if (!fs.existsSync(path_component)) continue;

      dependencies.add(path.resolve(path_component));

      const source = fs.readFileSync(path_component, "utf-8");
      const { html, css, script, module } = (
        await parseComponent({ source, filename })
      ).parsed;

      const source_formatted = prettier.format(source, { parser: "svelte" });
      const highlighted_code = Prism.highlight(
        source_formatted,
        Prism.languages.svelte,
        "svelte"
      );

      module.forEach((line) => script_module_unique_lines.add(line));
      script.forEach((line) => script_unique_lines.add(line));
      styles += css;

      let line_modified = "\n";

      if (noEval) {
        line_modified += `\`\`\`svelte\n${source_formatted}\`\`\``;
      } else {
        line_modified += `<div class="eval">\n${html}\n</div>\n`;
        line_modified += `<pre><code>{@html \`${highlighted_code}\`}</code></pre>\n\n`;
      }

      lines_code += line_modified + "\n\n";
    }
  }

  let content_script_module = `
<script context="module">
${[...script_module_unique_lines].join("\n")}
</script>\n`;
  let content_script = `<script>
${[...script_unique_lines].join("\n")}
</script>\n`;
  let content_style = `<style>\n${styles}\n</style>`;
  let code_eval = content_script_module + content_script + content_style;
  let code = noEval ? lines_code : code_eval + lines_code;

  return {
    code,
    dependencies: [...dependencies],
  };
};
