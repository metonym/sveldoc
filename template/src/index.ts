export { default } from "./Counter.svelte";
export { counter } from "./counter";
