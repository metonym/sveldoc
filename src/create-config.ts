/**
 * Future work
 *
 * 1. Read `package.json` to infer:
 *  - name
 *  - if `svelte-preprocess` is installed -> dynamically import it
 *  - which `@sveltejs/adapter-*` to use, if any -> dynamically import it
 *
 * 2. Should `mdsvex` be installed by default?
 */

import path from "path";
import { preprocessReadme, pluginReadme } from "./integrations";
import { mdsvex } from "mdsvex";
import preprocess from "svelte-preprocess";
import type { MdsvexOptions } from "mdsvex";
import type { Config as SvelteKitConfig, Adapter } from "@sveltejs/kit";
import type { AliasOptions, UserConfig as ViteConfig } from "vite";

interface CreateConfigOptions extends SvelteKitConfig {
  /**
   * Specify the package name.
   *
   * The default resolution path is assumed to be `src/lib`;
   * override it using `files.lib` or `kit.files.lib`.
   * @example "package-name"
   */
  name?: string;

  /**
   * Options passed to `mdsvex`.
   * @default
   * {
   *   extensions: [".md"],
   *   smartypants: false
   * }
   */
  mdsvexOptions?: MdsvexOptions;

  /**
   * Specify the SvelteKit adapter.
   * @example
   * import adapter from "@sveltejs/adapter-static";
   *
   * createConfig({
   *   adapter: adapter()
   * });
   */
  adapter?: Adapter;

  files?: {
    /**
     * @example "src"
     */
    lib?: string;

    /**
     * @example "demo"
     */
    routes?: string;

    /**
     * @example "demo/_app.html"
     */
    template?: string;

    serviceWorker?: string;
    assets?: string;
    hooks?: string;
  };
}

/**
 * Creates a `svelte.config.js` with sensible defaults.
 * Automatically includes `readme` preprocessor and
 * `sveldoc` vite plugin in the build process.
 *
 * Note: `svelte-preprocess` and `mdsvex` must
 * be installed as development dependencies.
 */
export const createConfig: (config: CreateConfigOptions) => Promise<SvelteKitConfig> = async (
  config
) => {
  const vite =
    typeof config?.kit?.vite === "function"
      ? ((await config.kit.vite()) as ViteConfig)
      : config?.kit?.vite;
  const alias: AliasOptions = {};
  const files = config?.files ?? config?.kit?.files;

  if (config?.name) {
    alias[config?.name] = path.resolve(files?.lib ?? "src/lib");
  }

  return {
    extensions: [".svelte", ".md"],
    preprocess: [
      preprocess(),
      preprocessReadme(),
      mdsvex({
        extensions: [".md"],
        smartypants: false,
        ...config?.mdsvexOptions,
      }),
    ],
    kit: {
      adapter: config?.adapter,
      files: config?.files,
      prerender: {
        default: true,
      },
      ...config?.kit,
      vite: {
        ...vite,
        plugins: [...(vite?.plugins ?? []), pluginReadme()],
        resolve: {
          ...vite?.resolve,
          alias: {
            ...alias,
            ...vite?.resolve?.alias,
          },
        },
        server: {
          ...vite?.server,
          watch: {
            ignored: ["!README.md"],
            ...vite?.server?.watch,
          },
          fs: {
            allow: [".."],
            ...vite?.server?.fs,
          },
        },
      },
    },
  };
};
