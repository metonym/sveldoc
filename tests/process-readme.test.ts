import { test, expect, describe } from "vitest";
import { processReadme } from "../src/process-readme";

const source = `
# heading

<!-- example-start -->
SKIP
<!-- example-end -->

<!-- example-start invalid.file -->
SKIP
<!-- example-end -->

<!-- example-start ./tests/__fixtures__/Button.svelte -->

Content


Another line

<!-- example-end -->

## heading 2

<!-- example-start tests/__fixtures__/Heading.svelte -->
<!-- example-end -->
    `;

describe("processReadme", () => {
  test("", async () => {
    const result_eval = await processReadme({ source, filename: "" });
    expect(result_eval.code).toMatchSnapshot();
    expect(result_eval.dependencies.length).toEqual(2);

    const result_no_eval = await processReadme({
      source,
      filename: "",
      noEval: true,
    });
    expect(result_no_eval.code).toMatchSnapshot();

    const result_re_eval = await processReadme({
      source: result_no_eval.code,
      filename: "",
    });
    expect(result_re_eval.code).toMatchSnapshot();
  });
});
