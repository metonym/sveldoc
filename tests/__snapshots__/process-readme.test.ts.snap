// Vitest Snapshot v1

exports[`processReadme >  1`] = `
"
<script context=\\"module\\">

</script>
<script>
import ComponentDefault from \\"./component\\";
import { ComponentNamed } from \\"./component\\";
import { Component as Alias } from \\"./component\\";
export let type = \\"button\\";
export let id = undefined;
</script>
<style>

  button {
    color: red;
  }

  h1 {
    font-size: 2rem;
  }

</style>
<h1 id=\\"heading\\">heading</h1>


<!-- example-start -->
<!-- example-end -->

<!-- example-start invalid.file -->
<!-- example-end -->

<!-- example-start ./tests/__fixtures__/Button.svelte -->

<div class=\\"eval\\">
<button {type}>
  <slot />
</button>
</div>
<pre><code>{@html \`<span class=\\"token tag\\"><span class=\\"token tag\\"><span class=\\"token punctuation\\">&lt;</span>script</span><span class=\\"token punctuation\\">></span></span><span class=\\"token script\\"><span class=\\"token language-javascript\\">
  <span class=\\"token keyword\\">import</span> ComponentDefault <span class=\\"token keyword\\">from</span> <span class=\\"token string\\">\\"./component\\"</span><span class=\\"token punctuation\\">;</span>
  <span class=\\"token keyword\\">import</span> <span class=\\"token punctuation\\">{</span> ComponentNamed <span class=\\"token punctuation\\">}</span> <span class=\\"token keyword\\">from</span> <span class=\\"token string\\">\\"./component\\"</span><span class=\\"token punctuation\\">;</span>
  <span class=\\"token keyword\\">import</span> <span class=\\"token punctuation\\">{</span> Component <span class=\\"token keyword\\">as</span> Alias <span class=\\"token punctuation\\">}</span> <span class=\\"token keyword\\">from</span> <span class=\\"token string\\">\\"./component\\"</span><span class=\\"token punctuation\\">;</span>

  <span class=\\"token keyword\\">export</span> <span class=\\"token keyword\\">let</span> type <span class=\\"token operator\\">=</span> <span class=\\"token string\\">\\"button\\"</span><span class=\\"token punctuation\\">;</span>
</span></span><span class=\\"token tag\\"><span class=\\"token tag\\"><span class=\\"token punctuation\\">&lt;/</span>script</span><span class=\\"token punctuation\\">></span></span>

<span class=\\"token tag\\"><span class=\\"token tag\\"><span class=\\"token punctuation\\">&lt;</span>button</span> <span class=\\"token language-javascript\\"><span class=\\"token punctuation\\">{</span>type<span class=\\"token punctuation\\">}</span></span><span class=\\"token punctuation\\">></span></span>
  <span class=\\"token tag\\"><span class=\\"token tag\\"><span class=\\"token punctuation\\">&lt;</span>slot</span> <span class=\\"token punctuation\\">/></span></span>
<span class=\\"token tag\\"><span class=\\"token tag\\"><span class=\\"token punctuation\\">&lt;/</span>button</span><span class=\\"token punctuation\\">></span></span>

<span class=\\"token tag\\"><span class=\\"token tag\\"><span class=\\"token punctuation\\">&lt;</span>style</span><span class=\\"token punctuation\\">></span></span><span class=\\"token style\\"><span class=\\"token language-css\\">
  <span class=\\"token selector\\">button</span> <span class=\\"token punctuation\\">{</span>
    <span class=\\"token property\\">color</span><span class=\\"token punctuation\\">:</span> red<span class=\\"token punctuation\\">;</span>
  <span class=\\"token punctuation\\">}</span>
</span></span><span class=\\"token tag\\"><span class=\\"token tag\\"><span class=\\"token punctuation\\">&lt;/</span>style</span><span class=\\"token punctuation\\">></span></span>
\`}</code></pre>



<!-- example-end -->

<h2 id=\\"heading-2\\">heading 2</h2>


<!-- example-start tests/__fixtures__/Heading.svelte -->

<div class=\\"eval\\">
<h1 {id}>
  <slot />
</h1>
</div>
<pre><code>{@html \`<span class=\\"token tag\\"><span class=\\"token tag\\"><span class=\\"token punctuation\\">&lt;</span>script</span><span class=\\"token punctuation\\">></span></span><span class=\\"token script\\"><span class=\\"token language-javascript\\">
  <span class=\\"token keyword\\">import</span> ComponentDefault <span class=\\"token keyword\\">from</span> <span class=\\"token string\\">\\"./component\\"</span><span class=\\"token punctuation\\">;</span>
  <span class=\\"token keyword\\">import</span> <span class=\\"token punctuation\\">{</span> ComponentNamed <span class=\\"token punctuation\\">}</span> <span class=\\"token keyword\\">from</span> <span class=\\"token string\\">\\"./component\\"</span><span class=\\"token punctuation\\">;</span>
  <span class=\\"token keyword\\">import</span> <span class=\\"token punctuation\\">{</span> Component <span class=\\"token keyword\\">as</span> Alias <span class=\\"token punctuation\\">}</span> <span class=\\"token keyword\\">from</span> <span class=\\"token string\\">\\"./component\\"</span><span class=\\"token punctuation\\">;</span>

  <span class=\\"token keyword\\">export</span> <span class=\\"token keyword\\">let</span> id <span class=\\"token operator\\">=</span> <span class=\\"token keyword\\">undefined</span><span class=\\"token punctuation\\">;</span>
</span></span><span class=\\"token tag\\"><span class=\\"token tag\\"><span class=\\"token punctuation\\">&lt;/</span>script</span><span class=\\"token punctuation\\">></span></span>

<span class=\\"token tag\\"><span class=\\"token tag\\"><span class=\\"token punctuation\\">&lt;</span>h1</span> <span class=\\"token language-javascript\\"><span class=\\"token punctuation\\">{</span>id<span class=\\"token punctuation\\">}</span></span><span class=\\"token punctuation\\">></span></span>
  <span class=\\"token tag\\"><span class=\\"token tag\\"><span class=\\"token punctuation\\">&lt;</span>slot</span> <span class=\\"token punctuation\\">/></span></span>
<span class=\\"token tag\\"><span class=\\"token tag\\"><span class=\\"token punctuation\\">&lt;/</span>h1</span><span class=\\"token punctuation\\">></span></span>

<span class=\\"token tag\\"><span class=\\"token tag\\"><span class=\\"token punctuation\\">&lt;</span>style</span><span class=\\"token punctuation\\">></span></span><span class=\\"token style\\"><span class=\\"token language-css\\">
  <span class=\\"token selector\\">h1</span> <span class=\\"token punctuation\\">{</span>
    <span class=\\"token property\\">font-size</span><span class=\\"token punctuation\\">:</span> 2rem<span class=\\"token punctuation\\">;</span>
  <span class=\\"token punctuation\\">}</span>
</span></span><span class=\\"token tag\\"><span class=\\"token tag\\"><span class=\\"token punctuation\\">&lt;/</span>style</span><span class=\\"token punctuation\\">></span></span>
\`}</code></pre>



<!-- example-end -->


"
`;

exports[`processReadme >  2`] = `
"
# heading

<!-- example-start -->
<!-- example-end -->

<!-- example-start invalid.file -->
<!-- example-end -->

<!-- example-start ./tests/__fixtures__/Button.svelte -->

\`\`\`svelte
<script>
  import ComponentDefault from \\"./component\\";
  import { ComponentNamed } from \\"./component\\";
  import { Component as Alias } from \\"./component\\";

  export let type = \\"button\\";
</script>

<button {type}>
  <slot />
</button>

<style>
  button {
    color: red;
  }
</style>
\`\`\`

<!-- example-end -->

## heading 2

<!-- example-start tests/__fixtures__/Heading.svelte -->

\`\`\`svelte
<script>
  import ComponentDefault from \\"./component\\";
  import { ComponentNamed } from \\"./component\\";
  import { Component as Alias } from \\"./component\\";

  export let id = undefined;
</script>

<h1 {id}>
  <slot />
</h1>

<style>
  h1 {
    font-size: 2rem;
  }
</style>
\`\`\`

<!-- example-end -->
    

"
`;

exports[`processReadme >  3`] = `
"
<script context=\\"module\\">

</script>
<script>
import ComponentDefault from \\"./component\\";
import { ComponentNamed } from \\"./component\\";
import { Component as Alias } from \\"./component\\";
export let type = \\"button\\";
export let id = undefined;
</script>
<style>

  button {
    color: red;
  }

  h1 {
    font-size: 2rem;
  }

</style>
<h1 id=\\"heading\\">heading</h1>


<!-- example-start -->
<!-- example-end -->

<!-- example-start invalid.file -->
<!-- example-end -->

<!-- example-start ./tests/__fixtures__/Button.svelte -->

<div class=\\"eval\\">
<button {type}>
  <slot />
</button>
</div>
<pre><code>{@html \`<span class=\\"token tag\\"><span class=\\"token tag\\"><span class=\\"token punctuation\\">&lt;</span>script</span><span class=\\"token punctuation\\">></span></span><span class=\\"token script\\"><span class=\\"token language-javascript\\">
  <span class=\\"token keyword\\">import</span> ComponentDefault <span class=\\"token keyword\\">from</span> <span class=\\"token string\\">\\"./component\\"</span><span class=\\"token punctuation\\">;</span>
  <span class=\\"token keyword\\">import</span> <span class=\\"token punctuation\\">{</span> ComponentNamed <span class=\\"token punctuation\\">}</span> <span class=\\"token keyword\\">from</span> <span class=\\"token string\\">\\"./component\\"</span><span class=\\"token punctuation\\">;</span>
  <span class=\\"token keyword\\">import</span> <span class=\\"token punctuation\\">{</span> Component <span class=\\"token keyword\\">as</span> Alias <span class=\\"token punctuation\\">}</span> <span class=\\"token keyword\\">from</span> <span class=\\"token string\\">\\"./component\\"</span><span class=\\"token punctuation\\">;</span>

  <span class=\\"token keyword\\">export</span> <span class=\\"token keyword\\">let</span> type <span class=\\"token operator\\">=</span> <span class=\\"token string\\">\\"button\\"</span><span class=\\"token punctuation\\">;</span>
</span></span><span class=\\"token tag\\"><span class=\\"token tag\\"><span class=\\"token punctuation\\">&lt;/</span>script</span><span class=\\"token punctuation\\">></span></span>

<span class=\\"token tag\\"><span class=\\"token tag\\"><span class=\\"token punctuation\\">&lt;</span>button</span> <span class=\\"token language-javascript\\"><span class=\\"token punctuation\\">{</span>type<span class=\\"token punctuation\\">}</span></span><span class=\\"token punctuation\\">></span></span>
  <span class=\\"token tag\\"><span class=\\"token tag\\"><span class=\\"token punctuation\\">&lt;</span>slot</span> <span class=\\"token punctuation\\">/></span></span>
<span class=\\"token tag\\"><span class=\\"token tag\\"><span class=\\"token punctuation\\">&lt;/</span>button</span><span class=\\"token punctuation\\">></span></span>

<span class=\\"token tag\\"><span class=\\"token tag\\"><span class=\\"token punctuation\\">&lt;</span>style</span><span class=\\"token punctuation\\">></span></span><span class=\\"token style\\"><span class=\\"token language-css\\">
  <span class=\\"token selector\\">button</span> <span class=\\"token punctuation\\">{</span>
    <span class=\\"token property\\">color</span><span class=\\"token punctuation\\">:</span> red<span class=\\"token punctuation\\">;</span>
  <span class=\\"token punctuation\\">}</span>
</span></span><span class=\\"token tag\\"><span class=\\"token tag\\"><span class=\\"token punctuation\\">&lt;/</span>style</span><span class=\\"token punctuation\\">></span></span>
\`}</code></pre>



<!-- example-end -->

<h2 id=\\"heading-2\\">heading 2</h2>


<!-- example-start tests/__fixtures__/Heading.svelte -->

<div class=\\"eval\\">
<h1 {id}>
  <slot />
</h1>
</div>
<pre><code>{@html \`<span class=\\"token tag\\"><span class=\\"token tag\\"><span class=\\"token punctuation\\">&lt;</span>script</span><span class=\\"token punctuation\\">></span></span><span class=\\"token script\\"><span class=\\"token language-javascript\\">
  <span class=\\"token keyword\\">import</span> ComponentDefault <span class=\\"token keyword\\">from</span> <span class=\\"token string\\">\\"./component\\"</span><span class=\\"token punctuation\\">;</span>
  <span class=\\"token keyword\\">import</span> <span class=\\"token punctuation\\">{</span> ComponentNamed <span class=\\"token punctuation\\">}</span> <span class=\\"token keyword\\">from</span> <span class=\\"token string\\">\\"./component\\"</span><span class=\\"token punctuation\\">;</span>
  <span class=\\"token keyword\\">import</span> <span class=\\"token punctuation\\">{</span> Component <span class=\\"token keyword\\">as</span> Alias <span class=\\"token punctuation\\">}</span> <span class=\\"token keyword\\">from</span> <span class=\\"token string\\">\\"./component\\"</span><span class=\\"token punctuation\\">;</span>

  <span class=\\"token keyword\\">export</span> <span class=\\"token keyword\\">let</span> id <span class=\\"token operator\\">=</span> <span class=\\"token keyword\\">undefined</span><span class=\\"token punctuation\\">;</span>
</span></span><span class=\\"token tag\\"><span class=\\"token tag\\"><span class=\\"token punctuation\\">&lt;/</span>script</span><span class=\\"token punctuation\\">></span></span>

<span class=\\"token tag\\"><span class=\\"token tag\\"><span class=\\"token punctuation\\">&lt;</span>h1</span> <span class=\\"token language-javascript\\"><span class=\\"token punctuation\\">{</span>id<span class=\\"token punctuation\\">}</span></span><span class=\\"token punctuation\\">></span></span>
  <span class=\\"token tag\\"><span class=\\"token tag\\"><span class=\\"token punctuation\\">&lt;</span>slot</span> <span class=\\"token punctuation\\">/></span></span>
<span class=\\"token tag\\"><span class=\\"token tag\\"><span class=\\"token punctuation\\">&lt;/</span>h1</span><span class=\\"token punctuation\\">></span></span>

<span class=\\"token tag\\"><span class=\\"token tag\\"><span class=\\"token punctuation\\">&lt;</span>style</span><span class=\\"token punctuation\\">></span></span><span class=\\"token style\\"><span class=\\"token language-css\\">
  <span class=\\"token selector\\">h1</span> <span class=\\"token punctuation\\">{</span>
    <span class=\\"token property\\">font-size</span><span class=\\"token punctuation\\">:</span> 2rem<span class=\\"token punctuation\\">;</span>
  <span class=\\"token punctuation\\">}</span>
</span></span><span class=\\"token tag\\"><span class=\\"token tag\\"><span class=\\"token punctuation\\">&lt;/</span>style</span><span class=\\"token punctuation\\">></span></span>
\`}</code></pre>



<!-- example-end -->




"
`;
